commitment_scheme/src/table_verifier.rs

```
@@ -0,0 +1,495 @@
+            &all_query_rows(data_queries, integrity_queries),
+            integrity_queries,
+        );
+
+        for query_loc in to_receive {
+            let mut channel = self.channel.borrow_mut();
+            let field_element = channel.recv_felts(1)?;
+            response.insert(query_loc, field_element[0]);
+        }
+
+        Ok(response)
+    }
+
+    /// Given indexed field elements, verify that these field elements are indeed the ones committed to
+    /// by the prover, against the commitment obtained by read_commitment().
+    #[allow(dead_code)]
+    fn verify_decommitment(&mut self, all_rows_data: &BTreeMap<RowCol, F>) -> Option<bool> {
+        let mut integrity_map: BTreeMap<usize, Vec<u8>> = BTreeMap::new();
+
+        let element_size = F::MODULUS_BIT_SIZE.div_ceil(8) as usize;
+
+        // Iterate over all rows in the map.
+        let mut all_rows_it = all_rows_data.iter().peekable();
+        while let Some((&row_col, _value)) = all_rows_it.peek() {
+            // Insert a new vector in the integrity_map with the current row number as the key.
+            let cur_row = row_col.get_row();
+            let mut row_data = vec![0u8; self.n_columns * element_size];
+            let mut pos = 0;
+
+            for _ in 0..self.n_columns {
+                if let Some((&row_col_next, field_element)) = all_rows_it.next() {
+                    assert_eq!(
+                        row_col_next.get_row(),
+                        cur_row,
+                        "Data skips to next row before finishing the current."
+                    );
+                    // Copy the field element bytes into the appropriate position in row_data.
+                    let field_bytes = field_element.into_bigint().to_bytes_be();
+                    row_data[pos..pos + element_size].copy_from_slice(&field_bytes);
+                    pos += element_size;
+                } else {
+                    panic!("Not enough columns in the map.");
```

Is this unreachable? Otherwise we should return an error.